tests/nexus_api_test.go

```diff
@@ -201,21 +351,221 @@ func (s *clientFunctionalSuite) TestNexusStartOperation_WithNamespaceAndTaskQueu
+			s.Equal(metrics.MetricUnit(metrics.Milliseconds), snap["nexus_latency"][0].Unit)
+		})
+	}
+}
+
+func (s *clientFunctionalSuite) TestNexusStartOperation_WithNamespaceAndTaskQueue_SupportsVersioning() {
+	ctx, cancel := context.WithCancel(NewContext())
+	defer cancel()
+	taskQueue := s.randomizeStr("task-queue")
+	err := s.sdkClient.UpdateWorkerBuildIdCompatibility(ctx, &sdkclient.UpdateWorkerBuildIdCompatibilityOptions{
+		TaskQueue: taskQueue,
+		Operation: &sdkclient.BuildIDOpAddNewIDInNewDefaultSet{BuildID: "old-build-id"},
+	})
+	s.Require().NoError(err)
+	err = s.sdkClient.UpdateWorkerBuildIdCompatibility(ctx, &sdkclient.UpdateWorkerBuildIdCompatibilityOptions{
+		TaskQueue: taskQueue,
+		Operation: &sdkclient.BuildIDOpAddNewIDInNewDefaultSet{BuildID: "new-build-id"},
+	})
+	s.Require().NoError(err)
+
+	u := fmt.Sprintf("http://%s/api/v1/namespaces/%s/task-queues/%s/dispatch-nexus-task", s.httpAPIAddress, s.namespace, taskQueue)
+	client, err := nexus.NewClient(nexus.ClientOptions{ServiceBaseURL: u})
+	s.Require().NoError(err)
+	// Versioned poller gets task
+	go s.versionedNexusTaskPoller(ctx, taskQueue, "new-build-id", nexusEchoHandler)
+
+	result, err := nexus.StartOperation(ctx, client, op, "input", nexus.StartOperationOptions{})
+	s.Require().NoError(err)
+	s.Require().Equal("input", result.Successful)
+
+	// Unversioned poller doesn't get a task
+	go s.nexusTaskPoller(ctx, taskQueue, nexusEchoHandler)
+	// Versioned poller gets task with wrong build ID
+	go s.versionedNexusTaskPoller(ctx, taskQueue, "old-build-id", nexusEchoHandler)
+
+	ctx, cancel = context.WithTimeout(ctx, time.Second*2)
+	defer cancel()
+	_, err = nexus.StartOperation(ctx, client, op, "input", nexus.StartOperationOptions{})
+	s.Require().ErrorIs(err, context.DeadlineExceeded)
+}
+
+func (s *clientFunctionalSuite) echoNexusTaskPoller(ctx context.Context, taskQueue string) {
+	s.nexusTaskPoller(ctx, taskQueue, nexusEchoHandler)
+}
+
+func (s *clientFunctionalSuite) nexusTaskPoller(ctx context.Context, taskQueue string, handler func(*workflowservice.PollNexusTaskQueueResponse) (*nexuspb.Response, *nexuspb.HandlerError)) {
+	s.versionedNexusTaskPoller(ctx, taskQueue, "", handler)
+}
+
+func (s *clientFunctionalSuite) versionedNexusTaskPoller(ctx context.Context, taskQueue, buildID string, handler func(*workflowservice.PollNexusTaskQueueResponse) (*nexuspb.Response, *nexuspb.HandlerError)) {
+	var vc *commonpb.WorkerVersionCapabilities
+
+	if buildID != "" {
+		vc = &commonpb.WorkerVersionCapabilities{
+			BuildId:       buildID,
+			UseVersioning: true,
+		}
+	}
+	res, err := s.testCluster.GetFrontendClient().PollNexusTaskQueue(ctx, &workflowservice.PollNexusTaskQueueRequest{
+		Namespace: s.namespace,
+		Identity:  uuid.NewString(),
+		TaskQueue: &taskqueuepb.TaskQueue{
+			Name: taskQueue,
+			Kind: enumspb.TASK_QUEUE_KIND_NORMAL,
+		},
+		WorkerVersionCapabilities: vc,
+	})
+	// We allow tests to leave pollers dangling and cancel the context.
```

Comment:
I'm not sure I understand this comment. Maybe you could elaborate here?

Reply:
I'll add to the code too.

If the context is already canceled, it's because the test is written in a way that it doesn't expect the poll to be unblocked.